### ISSUE TITLE
Dawn of War 3 assert in aco_opcode::p_parallelcopy

### ISSUE THREAD

**Issue by tarceri** [daniel-schuermann/mesa#21]
assert(instr->getDefinition(i).size() == 1);

I've also seen this happen in rise of the tomb raider when applying the out of tree GCM pass.

**Comment by tarceri** [comment]
Bug in ssa_elimination() ??? Assert is in block 41.

Before ssa elimination:
```
BB41
/* logical preds: BB40, / linear preds: BB40,  */
	p_logical_start
	p_logical_end
	 s2: %418:s[22-23],   b: %419:scc = s_andn2_b64 %390:s[22-23], %0:exec
	 s2: %0:exec = s_mov_b64 0x0
	p_cbranch_z %419:scc BB48, BB42
BB42
/* logical preds: / linear preds: BB41,  */
	p_branch BB44
BB43
/* logical preds: / linear preds: BB40,  */
	p_branch BB44
BB44
/* logical preds: / linear preds: BB42, BB43,  */
	 s2: %421:s[22-23] = s_mov_b64 %0:exec
	 s2: %672:exec = p_parallelcopy %416:s[28-29]
	p_cbranch_z %672:exec BB46, BB45
BB45
/* logical preds: BB40, / linear preds: BB44,  */
	p_logical_start
	p_logical_end
	p_branch BB47
BB46
/* logical preds: / linear preds: BB44,  */
	p_branch BB47
BB47
/* logical preds: BB45, / linear preds: BB45, BB46,  */
	 s2: %0:exec,   b: %422:scc = s_or_b64 %0:exec, %421:s[22-23]
	p_logical_start
	 s1: %392:s[20],  s1: %423:scc = s_add_i32 %391:s[20], 0x1
	 s1: %394:s[1],  s1: %424:scc = s_lshl_b32 %392:s[20], 0x1
	p_logical_end
	p_branch BB40
BB48
/* logical preds: BB41, / linear preds: BB41,  */
	 v1: %429:v[11] = p_phi %391:s[20]
	 v1: %428:v[12] = p_phi %393:s[1]
	 s2: %427:s[0-1] = p_linear_phi %671:s[21]
	 s2: %0:exec,   b: %425:scc = s_or_b64 %0:exec, %389:s[4-5]
	p_logical_start
	 s2: %430:vcc = v_cmp_lt_u32 0x0, %429:v[11]
	 s2: %431:s[0-1],   b: %432:scc = s_and_b64 %427:s[0-1], %430:vcc
	p_logical_end
	 s2: %433:s[2-3],   b: %434:scc,  s2: %435:exec = s_and_saveexec_b64 %431:s[0-1]
	 s2: %436:s[0-1],   b: %437:scc = s_andn2_b64 %433:s[2-3], %431:s[0-1]
	p_cbranch_z %435:exec BB50, BB49
```

After ssa elimination:

```
BB41
/* logical preds: BB40, / linear preds: BB40,  */
	p_logical_start
	 v1: %429:v[11],  v1: %428:v[12] = p_parallelcopy %391:s[20], %393:s[1]
	p_logical_end
	 s2: %418:s[22-23],   b: %419:scc = s_andn2_b64 %390:s[22-23], %0:exec
	 s2: %0:exec = s_mov_b64 0x0
	 s2: %427:s[0-1] = p_parallelcopy %671:s[21]
	p_cbranch_z %419:scc BB48, BB47
BB42
/* logical preds: / linear preds:  */
BB43
/* logical preds: / linear preds:  */
BB44
/* logical preds: / linear preds:  */
BB45
/* logical preds: BB40, / linear preds:  */
BB46
/* logical preds: / linear preds:  */
BB47
/* logical preds: BB45, / linear preds: BB41, BB40,  */
	 s2: %0:exec,   b: %422:scc = s_or_b64 %0:exec, %416:s[28-29]
	p_logical_start
	 s1: %392:s[20],  s1: %423:scc = s_add_i32 %391:s[20], 0x1
	 s1: %394:s[1],  s1: %424:scc = s_lshl_b32 %392:s[20], 0x1
	p_logical_end
	p_branch BB40
BB48
/* logical preds: BB41, / linear preds: BB41,  */
	 v1: %429:v[11] = p_phi %391:s[20]
	 v1: %428:v[12] = p_phi %393:s[1]
	 s2: %427:s[0-1] = p_linear_phi %671:s[21]
	 s2: %0:exec,   b: %425:scc = s_or_b64 %0:exec, %389:s[4-5]
	p_logical_start
	 s2: %430:vcc = v_cmp_lt_u32 0x0, %429:v[11]
	 s2: %431:s[0-1],   b: %432:scc = s_and_b64 %427:s[0-1], %430:vcc
	p_logical_end
	 s2: %433:s[2-3],   b: %434:scc,  s2: %435:exec = s_and_saveexec_b64 %431:s[0-1]
	 s2: %436:s[0-1],   b: %437:scc = s_andn2_b64 %433:s[2-3], %431:s[0-1]
	p_cbranch_z %435:exec BB51, BB49
```

**Comment by pendingchaos** [comment]
It's an issue with instruction selection. https://github.com/daniel-schuermann/mesa/blob/backend/src/amd/compiler/aco_instruction_selection.cpp#L5018-L5019 is creating a linear phi with a s2 destination and a s1 source (`s2: %427:s[0-1] = p_linear_phi %671:s[21]`).

**Comment by pendingchaos** [comment]
Should be fixed in 5ecdeaec043967c8ac6345481da2181acf9113eb.
